UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.FulfilmentService/Services/FulfilmentCallBackService.cs

```diff
@@ -0,0 +1,95 @@
+                                         ICallBackClient callBackClient,
+                                         IOptions<FileShareServiceConfiguration> fileShareServiceConfig,
+                                         ILogger<FulfilmentCallBackService> logger)
+        {
+            this.essCallBackConfiguration = essCallBackConfiguration;
+            this.callBackClient = callBackClient;
+            this.fileShareServiceConfig = fileShareServiceConfig;
+            this.logger = logger;
+        }
+
+        public async Task<bool> SendCallBackResponse(SalesCatalogueProductResponse salesCatalogueProductResponse, SalesCatalogueServiceResponseQueueMessage scsResponseQueueMessage)
+        {
+            if (!string.IsNullOrWhiteSpace(scsResponseQueueMessage.CallbackUri))
+            {
+                ExchangeSetResponse exchangeSetResponse = new ExchangeSetResponse()
+                {
+                    Links = new Links()
+                    {
+                        ExchangeSetBatchStatusUri = new LinkSetBatchStatusUri { Href = $"{fileShareServiceConfig.Value.BaseUrl}/batch/{scsResponseQueueMessage.BatchId}" },
+                        ExchangeSetFileUri = new LinkSetFileUri { Href = $"{fileShareServiceConfig.Value.BaseUrl}/batch/{scsResponseQueueMessage.BatchId}/files/{fileShareServiceConfig.Value.ExchangeSetFileName}" }
+                    },
+                    ExchangeSetUrlExpiryDateTime = Convert.ToDateTime(scsResponseQueueMessage.ExchangeSetUrlExpiryDate).ToUniversalTime(),
+                    RequestedProductCount = salesCatalogueProductResponse.ProductCounts.RequestedProductCount.Value,
+                    ExchangeSetCellCount = salesCatalogueProductResponse.ProductCounts.ReturnedProductCount.Value,
+                    RequestedProductsAlreadyUpToDateCount = salesCatalogueProductResponse.ProductCounts.RequestedProductsAlreadyUpToDateCount.Value,
+                    RequestedProductsNotInExchangeSet = GetRequestedProductsNotInExchangeSet(salesCatalogueProductResponse)
+                };
+
+                CallBackResponse callBackResponse = new CallBackResponse()
+                {
+                    Specversion = essCallBackConfiguration.Value.SpecVersion,
+                    Type = essCallBackConfiguration.Value.Type,
+                    Source = essCallBackConfiguration.Value.Source,
+                    Id = Guid.NewGuid().ToString(),
+                    Time = DateTime.UtcNow.ToString("yyyy-MM-ddTHH:mm:ssZ", CultureInfo.InvariantCulture),
+                    Subject = essCallBackConfiguration.Value.Subject,
+                    DataContentType = "application/json",
+                    Data = exchangeSetResponse
+                };
+
+                string payloadJson = JsonConvert.SerializeObject(callBackResponse);
+
+                await callBackClient.CallBackApi(HttpMethod.Post, payloadJson, scsResponseQueueMessage.CallbackUri);
```

Comment:
This feels like a point of failure. What behaviour would we see/expect if we had an invalid Uri or the client doesn't respond etc

Is this an area of the code where we should specifically manage that behaviour i.e. log the issue and continue or do we allow it to fail?

Sorry I'm not that close to the rest of the process to have a definitive answer.

Reply:
As discussed in last refinement,validating uri is not part of this pbi.
However there is a separate PBI 9043(Re-visit call-back URI validation) in the backlog to be discussed later.
As per the ESS API spec the service will ignore all response from the callback and hence we are not validating response